File: src/com/campusnet/simulator/NetworkSimulator.java
package com.campusnet.simulator;

import com.campusnet.model.*;
import com.campusnet.protocol.*;

import java.util.*;
import java.util.concurrent.*;

public class NetworkSimulator {
    private final List<Device> devices = new CopyOnWriteArrayList<>();
    private final List<NetworkLink> links = new CopyOnWriteArrayList<>();
    private final List<Packet> activeInFlightPackets = new CopyOnWriteArrayList<>();
    private final List<Packet> packetHistory = new CopyOnWriteArrayList<>();
    private final List<SimulationListener> listeners = new CopyOnWriteArrayList<>();

    private final NetworkMetrics metrics = new NetworkMetrics();
    private final NatEngine natEngine = new NatEngine();
    private final FirewallEngine firewallEngine = new FirewallEngine();
    private final DnsService dnsService = new DnsService();

    private volatile boolean isRunning = true;
    private volatile boolean isPaused = false;
    private volatile double simulationSpeed = 1.0; // 0.2x to 10.0x
    private ScheduledExecutorService executorService;
    private ScheduledFuture<?> simulationTask;
    private ScheduledFuture<?> metricsTask;

    public NetworkSimulator() {
        startSimulationLoop();
    }

    public void addListener(SimulationListener listener) {
        listeners.add(listener);
    }

    public void removeListener(SimulationListener listener) {
        listeners.remove(listener);
    }

    public void addDevice(Device device) {
        devices.add(device);
        notifyTopologyChanged();
    }

    public void addLink(NetworkLink link) {
        links.add(link);
        notifyTopologyChanged();
    }

    public List<Device> getDevices() {
        return devices;
    }

    public List<NetworkLink> getLinks() {
        return links;
    }

    public List<Packet> getActiveInFlightPackets() {
        return activeInFlightPackets;
    }

    public List<Packet> getPacketHistory() {
        return packetHistory;
    }

    public NetworkMetrics getMetrics() {
        return metrics;
    }

    public NatEngine getNatEngine() {
        return natEngine;
    }

    public FirewallEngine getFirewallEngine() {
        return firewallEngine;
    }

    public DnsService getDnsService() {
        return dnsService;
    }

    public Device findDeviceByName(String name) {
        for (Device d : devices) {
            if (d.getName().equalsIgnoreCase(name)) {
                return d;
            }
        }
        return null;
    }

    public Device findDeviceByIp(String ip) {
        for (Device d : devices) {
            if (d.hasIp(ip)) {
                return d;
            }
        }
        return null;
    }

    public NetworkLink findLinkBetween(Device a, Device b) {
        for (NetworkLink link : links) {
            Device devA = link.getInterfaceA() != null ? link.getInterfaceA().getOwnerDevice() : null;
            Device devB = link.getInterfaceB() != null ? link.getInterfaceB().getOwnerDevice() : null;
            if ((devA == a && devB == b) || (devA == b && devB == a)) {
                return link;
            }
        }
        return null;
    }

    public void recalculateRouting() {
        RoutingEngine.computeAllRoutes(devices, links);
        notifyTopologyChanged();
        log("Dynamic routing tables recomputed (Dijkstra SPF).");
    }

    public void toggleLinkState(NetworkLink link) {
        link.setUp(!link.isUp());
        log("Link " + link.getId() + " toggled: " + (link.isUp() ? "UP" : "DOWN"));
        recalculateRouting();
    }

    /**
     * Injects a packet into the network starting from source device
     */
    public boolean sendPacket(Packet packet) {
        Device srcDev = findDeviceByIp(packet.getSourceIp());
        if (srcDev == null) {
            srcDev = findDeviceByName(packet.getSourceIp());
        }
        if (srcDev == null) {
            dropPacket(packet, "Source device not found for IP: " + packet.getSourceIp());
            return false;
        }

        packet.addHop(srcDev.getName());
        packet.setCurrentDevice(srcDev);
        packet.setStatus(Packet.Status.IN_TRANSIT);

        metrics.recordPacketSent(packet.getProtocol(), packet.getPacketSize());
        packetHistory.add(packet);
        if (packetHistory.size() > 2000) {
            packetHistory.remove(0);
        }

        for (SimulationListener listener : listeners) {
            listener.onPacketCreated(packet);
        }

        log(String.format("Packet %s [%s] injected at %s (%s -> %s)",
                packet.getId(), packet.getProtocol().getCode(), srcDev.getName(), packet.getSourceIp(), packet.getDestIp()));

        // Forward from initial source
        routeFromDevice(srcDev, packet);
        return true;
    }

    private void routeFromDevice(Device currentDev, Packet packet) {
        if (!currentDev.isOnline()) {
            dropPacket(packet, "Device " + currentDev.getName() + " is offline");
            return;
        }

        // 1. Check if packet arrived at destination
        if (currentDev.hasIp(packet.getDestIp())) {
            deliverPacket(currentDev, packet);
            return;
        }

        // 2. Check TTL
        if (packet.getTtl() <= 0) {
            dropPacket(packet, "TTL Expired in transit at " + currentDev.getName());
            return;
        }

        // 3. Device specific routing / forwarding
        if (currentDev instanceof SwitchDevice sw) {
            handleSwitchForward(sw, packet);
        } else if (currentDev instanceof RouterDevice router) {
            handleRouterForward(router, packet);
        } else if (currentDev instanceof HostDevice host) {
            handleHostForward(host, packet);
        }
    }

    private void handleHostForward(HostDevice host, Packet packet) {
        NetworkInterface iface = host.getPrimaryInterface();
        if (iface == null || !iface.isUp()) {
            dropPacket(packet, "Host " + host.getName() + " interface is DOWN");
            return;
        }

        NetworkLink link = iface.getConnectedLink();
        if (link == null || !link.isUp()) {
            dropPacket(packet, "No physical link on " + host.getName());
            return;
        }

        Device nextDev = link.getOtherDevice(host);
        if (nextDev == null) {
            dropPacket(packet, "Cable unplugged at " + host.getName());
            return;
        }

        // Learn ARP if direct subnet
        if (iface.isInSameSubnet(packet.getDestIp())) {
            packet.setDestMac("FF:FF:FF:FF:FF:00"); // Direct LAN frame
        } else {
            // Need default gateway
            if (host.getDefaultGateway() == null) {
                dropPacket(packet, "No default gateway configured on " + host.getName());
                return;
            }
        }

        transmitOnLink(link, host, nextDev, packet);
    }

    private void handleSwitchForward(SwitchDevice sw, Packet packet) {
        // Learn source MAC on incoming link/port
        NetworkLink inLink = packet.getCurrentLink();
        if (inLink != null) {
            NetworkInterface inIface = (inLink.getInterfaceA().getOwnerDevice() == sw) ?
                    inLink.getInterfaceA() : inLink.getInterfaceB();
            if (inIface != null && packet.getSourceMac() != null) {
                sw.learnMac(packet.getSourceMac(), inIface.getName(), packet.getVlanId());
            }
        }

        // Find destination host / router link
        Device destDev = findDeviceByIp(packet.getDestIp());
        NetworkLink targetLink = null;
        Device targetNextHop = null;

        // 1. Check direct connection to switch
        if (destDev != null) {
            targetLink = findLinkBetween(sw, destDev);
            if (targetLink != null && targetLink.isUp()) {
                targetNextHop = destDev;
            }
        }

        // 2. If not directly attached, find uplink Router connected to switch
        if (targetLink == null) {
            for (NetworkInterface swIface : sw.getInterfaces()) {
                NetworkLink link = swIface.getConnectedLink();
                if (link != null && link.isUp() && link != inLink) {
                    Device other = link.getOtherDevice(sw);
                    if (other instanceof RouterDevice) {
                        targetLink = link;
                        targetNextHop = other;
                        break;
                    }
                }
            }
        }

        // 3. Fallback: flood out of another port if not inLink
        if (targetLink == null) {
            for (NetworkInterface swIface : sw.getInterfaces()) {
                NetworkLink link = swIface.getConnectedLink();
                if (link != null && link.isUp() && link != inLink) {
                    Device other = link.getOtherDevice(sw);
                    if (other != null) {
                        targetLink = link;
                        targetNextHop = other;
                        break;
                    }
                }
            }
        }

        if (targetLink == null || !targetLink.isUp() || targetNextHop == null) {
            dropPacket(packet, "Switch " + sw.getName() + " forwarding failed (no active port)");
            return;
        }

        transmitOnLink(targetLink, sw, targetNextHop, packet);
    }

    private void handleRouterForward(RouterDevice router, Packet packet) {
        packet.decrementTtl();

        // 1. Firewall ACL evaluation
        if (router.isFirewallEnabled()) {
            FirewallEngine.FirewallResult fwResult = firewallEngine.evaluate(packet);
            if (!fwResult.isAllowed()) {
                dropPacket(packet, "Firewall Rule #" +
                        (fwResult.getMatchedRule() != null ? fwResult.getMatchedRule().getRuleNumber() : "Default") +
                        " BLOCKED packet at " + router.getName());
                return;
            }
        }

        // 2. Inbound NAT check (WAN to LAN response)
        if (router.isNatEnabled()) {
            natEngine.applyInboundNat(packet);
        }

        // 3. Routing Table Lookup
        RouteEntry route = router.findBestRoute(packet.getDestIp());
        if (route == null) {
            dropPacket(packet, "No route to destination " + packet.getDestIp() + " at " + router.getName());
            return;
        }

        NetworkInterface outIface = router.getInterfaceByName(route.getOutgoingInterface());
        if (outIface == null || !outIface.isUp()) {
            dropPacket(packet, "Outgoing interface " + route.getOutgoingInterface() + " is DOWN on " + router.getName());
            return;
        }

        NetworkLink link = outIface.getConnectedLink();
        if (link == null || !link.isUp()) {
            dropPacket(packet, "Outgoing link is DOWN on " + router.getName());
            return;
        }

        // 4. Outbound NAT check (LAN to WAN translation)
        if (router.isNatEnabled() && link.getAreaType() == AreaType.WAN) {
            natEngine.applyOutboundNat(packet, outIface.getIpAddress());
        }

        Device nextDev = link.getOtherDevice(router);
        if (nextDev == null) {
            dropPacket(packet, "No neighbor device on interface " + outIface.getName());
            return;
        }

        transmitOnLink(link, router, nextDev, packet);
    }

    private void transmitOnLink(NetworkLink link, Device fromDev, Device toDev, Packet packet) {
        // Check link packet loss simulation
        if (link.getPacketLossRate() > 0 && Math.random() < link.getPacketLossRate()) {
            dropPacket(packet, String.format("Packet lost on link %s (loss rate %.1f%%)", link.getId(), link.getPacketLossRate() * 100));
            return;
        }

        packet.setCurrentLink(link);
        packet.setCurrentDevice(fromDev);
        packet.setProgressOnLink(0.0);

        if (!activeInFlightPackets.contains(packet)) {
            activeInFlightPackets.add(packet);
        }

        log(String.format("Link Transmission: %s -> %s over %s [%s, latency=%.1fms]",
                fromDev.getName(), toDev.getName(), link.getId(), link.getAreaType().name(), link.getLatencyMs()));
    }

    private void deliverPacket(Device destDev, Packet packet) {
        packet.setStatus(Packet.Status.DELIVERED);
        packet.setCurrentDevice(destDev);
        packet.setCurrentLink(null);
        packet.setProgressOnLink(1.0);
        activeInFlightPackets.remove(packet);

        long latency = System.currentTimeMillis() - packet.getCreationTime();
        metrics.recordPacketDelivered(latency);

        packet.addHop(destDev.getName());

        for (SimulationListener listener : listeners) {
            listener.onPacketDelivered(packet);
        }

        log(String.format("SUCCESS: Packet %s DELIVERED at %s (%s). End-to-End Latency: %d ms, Hops: %s",
                packet.getId(), destDev.getName(), destDev.getPrimaryIp(), latency, String.join(" -> ", packet.getHopPath())));

        // Handle Application Layer Automated Replies
        processServerReply(destDev, packet);
    }

    private void dropPacket(Packet packet, String reason) {
        packet.setDropReason(reason);
        packet.setStatus(Packet.Status.DROPPED);
        activeInFlightPackets.remove(packet);
        metrics.recordPacketDropped();

        for (SimulationListener listener : listeners) {
            listener.onPacketDropped(packet, reason);
        }

        log(String.format("DROP: Packet %s DROPPED. Reason: %s", packet.getId(), reason));
    }

    private void processServerReply(Device destDev, Packet packet) {
        // Automatic ICMP Ping Echo Reply
        if (packet.getProtocol() == ProtocolType.ICMP && packet.getPayload().contains("Echo Request")) {
            Packet reply = Packet.createPingReply(packet.getDestIp(), packet.getSourceIp(), packet.getPingSequence());
            reply.setSourceMac(destDev.getPrimaryMac());
            // Delay slightly for realism
            CompletableFuture.delayedExecutor(100, TimeUnit.MILLISECONDS).execute(() -> sendPacket(reply));
        }
        // Automatic HTTP Server Reply
        else if (packet.getProtocol() == ProtocolType.HTTP && destDev instanceof ServerDevice server && server.isHttpServiceRunning()) {
            if ("GET".equalsIgnoreCase(packet.getHttpMethod())) {
                String page = server.getWebPage(packet.getHttpUrl() != null ? packet.getHttpUrl() : "/");
                Packet httpReply = Packet.createHttpResponse(packet.getDestIp(), packet.getSourceIp(), 200, page);
                CompletableFuture.delayedExecutor(150, TimeUnit.MILLISECONDS).execute(() -> sendPacket(httpReply));
            }
        }
        // Automatic DNS Server Reply
        else if (packet.getProtocol() == ProtocolType.DNS && destDev instanceof ServerDevice server && server.isDnsServiceRunning()) {
            String domain = packet.getDnsQuery();
            String resolved = dnsService.resolve(domain);
            if (resolved != null) {
                Packet dnsReply = Packet.createDnsResponse(packet.getDestIp(), packet.getSourceIp(), domain, resolved);
                CompletableFuture.delayedExecutor(80, TimeUnit.MILLISECONDS).execute(() -> sendPacket(dnsReply));
            }
        }
    }

    private void startSimulationLoop() {
        executorService = Executors.newScheduledThreadPool(2);

        // Animation and packet advancement tick (every 30ms)
        simulationTask = executorService.scheduleAtFixedRate(() -> {
            if (!isRunning || isPaused) return;
            try {
                tickPackets();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }, 0, 30, TimeUnit.MILLISECONDS);

        // Metrics sampler tick (every 1000ms)
        metricsTask = executorService.scheduleAtFixedRate(() -> {
            try {
                metrics.tickSample();
                for (SimulationListener listener : listeners) {
                    listener.onStatsUpdated(metrics);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }, 1, 1, TimeUnit.SECONDS);
    }

    private void tickPackets() {
        for (Packet packet : activeInFlightPackets) {
            NetworkLink link = packet.getCurrentLink();
            if (link == null) continue;

            double speedFactor = simulationSpeed;
            // Advance progress based on link latency (higher latency = slower packet movement)
            double step = (0.05 * speedFactor) / Math.max(link.getLatencyMs() / 10.0, 0.5);
            double newProgress = packet.getProgressOnLink() + step;

            if (newProgress >= 1.0) {
                // Packet arrived at end of link
                Device fromDev = packet.getCurrentDevice();
                Device nextDev = link.getOtherDevice(fromDev);
                packet.setProgressOnLink(1.0);
                activeInFlightPackets.remove(packet);

                if (nextDev != null) {
                    packet.addHop(nextDev.getName());
                    packet.setCurrentDevice(nextDev);
                    routeFromDevice(nextDev, packet);
                } else {
                    dropPacket(packet, "Link leads to nowhere");
                }
            } else {
                packet.setProgressOnLink(newProgress);
                for (SimulationListener listener : listeners) {
                    listener.onPacketMoved(packet, packet.getCurrentDevice() != null ? packet.getCurrentDevice().getName() : "?",
                            link.getOtherDevice(packet.getCurrentDevice()) != null ? link.getOtherDevice(packet.getCurrentDevice()).getName() : "?",
                            newProgress);
                }
            }
        }
    }

    public void play() {
        this.isPaused = false;
        log("Simulation resumed.");
    }

    public void pause() {
        this.isPaused = true;
        log("Simulation paused.");
    }

    public void step() {
        if (!isPaused) pause();
        tickPackets();
        log("Simulation single-step executed.");
    }

    public void setSimulationSpeed(double speed) {
        this.simulationSpeed = Math.max(0.1, Math.min(10.0, speed));
        log(String.format("Simulation speed set to %.1fx", this.simulationSpeed));
    }

    public double getSimulationSpeed() {
        return simulationSpeed;
    }

    public boolean isPaused() {
        return isPaused;
    }

    public void log(String message) {
        for (SimulationListener listener : listeners) {
            listener.onLogMessage(message);
        }
    }

    private void notifyTopologyChanged() {
        for (SimulationListener listener : listeners) {
            listener.onTopologyChanged();
        }
    }

    public void reset() {
        activeInFlightPackets.clear();
        packetHistory.clear();
        metrics.reset();
        for (Device d : devices) {
            d.getArpTable().clear();
            if (d instanceof SwitchDevice sw) {
                sw.clearMacTable();
            }
        }
        recalculateRouting();
        log("Simulation state reset.");
    }

    public void shutdown() {
        isRunning = false;
        if (simulationTask != null) simulationTask.cancel(true);
        if (metricsTask != null) metricsTask.cancel(true);
        if (executorService != null) executorService.shutdown();
    }
}

File: js/simulator/NetworkSimulator.js
(function(exports) {
    const ProtocolType = (typeof require !== 'undefined') ? require('../models/ProtocolType').ProtocolType : exports.ProtocolType;
    const Packet = (typeof require !== 'undefined') ? require('../models/Packet').Packet : exports.Packet;
    const RoutingEngine = (typeof require !== 'undefined') ? require('../protocols/RoutingEngine').RoutingEngine : exports.RoutingEngine;
    const NatEngine = (typeof require !== 'undefined') ? require('../protocols/NatEngine').NatEngine : exports.NatEngine;
    const FirewallEngine = (typeof require !== 'undefined') ? require('../protocols/FirewallEngine').FirewallEngine : exports.FirewallEngine;
    const DnsService = (typeof require !== 'undefined') ? require('../protocols/DnsService').DnsService : exports.DnsService;
    const NetworkMetrics = (typeof require !== 'undefined') ? require('./NetworkMetrics').NetworkMetrics : exports.NetworkMetrics;

    class NetworkSimulator {
        constructor() {
            this.devices = [];
            this.links = [];
            this.activeInFlightPackets = [];
            this.packetHistory = [];
            this.listeners = [];

            this.metrics = new NetworkMetrics();
            this.natEngine = new NatEngine();
            this.firewallEngine = new FirewallEngine();
            this.dnsService = new DnsService();

            this.isRunning = true;
            this.isPaused = false;
            this.simulationSpeed = 1.0; // 0.2x to 5.0x

            this.startLoop();
        }

        addListener(listener) {
            this.listeners.push(listener);
        }

        removeListener(listener) {
            this.listeners = this.listeners.filter(l => l !== listener);
        }

        addDevice(device) {
            this.devices.push(device);
            this.notify('onTopologyChanged');
        }

        addLink(link) {
            this.links.push(link);
            this.notify('onTopologyChanged');
        }

        findDeviceByName(name) {
            if (!name) return null;
            return this.devices.find(d => d.name.toLowerCase() === name.toLowerCase()) || null;
        }

        findDeviceByIp(ip) {
            if (!ip) return null;
            return this.devices.find(d => d.hasIp(ip)) || null;
        }

        findLinkBetween(a, b) {
            for (const link of this.links) {
                const devA = link.interfaceA ? link.interfaceA.ownerDevice : null;
                const devB = link.interfaceB ? link.interfaceB.ownerDevice : null;
                if ((devA === a && devB === b) || (devA === b && devB === a)) {
                    return link;
                }
            }
            return null;
        }

        recalculateRouting() {
            RoutingEngine.computeAllRoutes(this.devices, this.links);
            this.notify('onTopologyChanged');
            this.log('Dynamic routing tables recomputed (Dijkstra SPF).');
        }

        toggleLinkState(link) {
            link.isUp = !link.isUp;
            this.log(`Link ${link.id} toggled: ${link.isUp ? 'UP' : 'DOWN'}`);
            this.recalculateRouting();
        }

        sendPacket(packet) {
            let srcDev = this.findDeviceByIp(packet.sourceIp);
            if (!srcDev) {
                srcDev = this.findDeviceByName(packet.sourceIp);
            }
            if (!srcDev) {
                this.dropPacket(packet, `Source device not found for IP: ${packet.sourceIp}`);
                return false;
            }

            packet.addHop(srcDev.name);
            packet.currentDevice = srcDev;
            packet.status = 'IN_TRANSIT';

            this.metrics.recordPacketSent(packet.protocol, packet.getPacketSize());
            this.packetHistory.push(packet);
            if (this.packetHistory.length > 2000) {
                this.packetHistory.shift();
            }

            this.notify('onPacketCreated', packet);
            this.log(`Packet ${packet.id} [${packet.protocol ? packet.protocol.code : 'DATA'}] injected at ${srcDev.name} (${packet.sourceIp} -> ${packet.destIp})`);

            this.routeFromDevice(srcDev, packet);
            return true;
        }

        routeFromDevice(currentDev, packet) {
            if (!currentDev.isOnline) {
                this.dropPacket(packet, `Device ${currentDev.name} is offline`);
                return;
            }

            // 1. Destination reached?
            if (currentDev.hasIp(packet.destIp)) {
                this.deliverPacket(currentDev, packet);
                return;
            }

            // 2. TTL Check
            if (packet.ttl <= 0) {
                this.dropPacket(packet, `TTL Expired in transit at ${currentDev.name}`);
                return;
            }

            // 3. Device specific forwarding
            const type = currentDev.getDeviceType();
            if (type === 'L2 Switch') {
                this.handleSwitchForward(currentDev, packet);
            } else if (type === 'L3 Router') {
                this.handleRouterForward(currentDev, packet);
            } else {
                this.handleHostForward(currentDev, packet);
            }
        }

        handleHostForward(host, packet) {
            const iface = host.getPrimaryInterface();
            if (!iface || !iface.isUp) {
                this.dropPacket(packet, `Host ${host.name} interface is DOWN`);
                return;
            }

            const link = iface.connectedLink;
            if (!link || !link.isUp) {
                this.dropPacket(packet, `No physical link on ${host.name}`);
                return;
            }

            const nextDev = link.getOtherDevice(host);
            if (!nextDev) {
                this.dropPacket(packet, `Cable unplugged at ${host.name}`);
                return;
            }

            if (!iface.isInSameSubnet(packet.destIp)) {
                if (!host.defaultGateway) {
                    this.dropPacket(packet, `No default gateway configured on ${host.name}`);
                    return;
                }
            }

            this.transmitOnLink(link, host, nextDev, packet);
        }

        handleSwitchForward(sw, packet) {
            const inLink = packet.currentLink;
            if (inLink) {
                const inIface = (inLink.interfaceA.ownerDevice === sw) ? inLink.interfaceA : inLink.interfaceB;
                if (inIface && packet.sourceMac) {
                    sw.learnMac(packet.sourceMac, inIface.name, packet.vlanId);
                }
            }

            const destDev = this.findDeviceByIp(packet.destIp);
            let targetLink = null;
            let targetNextHop = null;

            // 1. Check direct connection to switch
            if (destDev) {
                targetLink = this.findLinkBetween(sw, destDev);
                if (targetLink && targetLink.isUp) {
                    targetNextHop = destDev;
                }
            }

            // 2. Uplink to Router
            if (!targetLink) {
                for (const swIface of sw.interfaces) {
                    const link = swIface.connectedLink;
                    if (link && link.isUp && link !== inLink) {
                        const other = link.getOtherDevice(sw);
                        if (other && other.getDeviceType() === 'L3 Router') {
                            targetLink = link;
                            targetNextHop = other;
                            break;
                        }
                    }
                }
            }

            // 3. Flooding fallback
            if (!targetLink) {
                for (const swIface of sw.interfaces) {
                    const link = swIface.connectedLink;
                    if (link && link.isUp && link !== inLink) {
                        const other = link.getOtherDevice(sw);
                        if (other) {
                            targetLink = link;
                            targetNextHop = other;
                            break;
                        }
                    }
                }
            }

            if (!targetLink || !targetLink.isUp || !targetNextHop) {
                this.dropPacket(packet, `Switch ${sw.name} forwarding failed (no active port)`);
                return;
            }

            this.transmitOnLink(targetLink, sw, targetNextHop, packet);
        }

        handleRouterForward(router, packet) {
            packet.decrementTtl();

            // 1. Firewall ACL
            if (router.firewallEnabled) {
                const fwResult = this.firewallEngine.evaluate(packet);
                if (!fwResult.allowed) {
                    this.dropPacket(packet, `Firewall Rule #${fwResult.matchedRule ? fwResult.matchedRule.ruleNumber : 'Default'} BLOCKED packet at ${router.name}`);
                    return;
                }
            }

            // 2. Inbound NAT check (WAN to LAN response)
            if (router.natEnabled) {
                this.natEngine.applyInboundNat(packet);
            }

            // 3. Routing Table Lookup
            const route = router.findBestRoute(packet.destIp);
            if (!route) {
                this.dropPacket(packet, `No route to destination ${packet.destIp} at ${router.name}`);
                return;
            }

            const outIface = router.getInterfaceByName(route.outgoingInterface);
            if (!outIface || !outIface.isUp) {
                this.dropPacket(packet, `Outgoing interface ${route.outgoingInterface} is DOWN on ${router.name}`);
                return;
            }

            const link = outIface.connectedLink;
            if (!link || !link.isUp) {
                this.dropPacket(packet, `Outgoing link is DOWN on ${router.name}`);
                return;
            }

            // 4. Outbound NAT check (LAN to WAN translation)
            if (router.natEnabled && link.areaType && link.areaType.id === 'WAN') {
                this.natEngine.applyOutboundNat(packet, outIface.ipAddress);
            }

            const nextDev = link.getOtherDevice(router);
            if (!nextDev) {
                this.dropPacket(packet, `No neighbor device on interface ${outIface.name}`);
                return;
            }

            this.transmitOnLink(link, router, nextDev, packet);
        }

        transmitOnLink(link, fromDev, toDev, packet) {
            if (link.packetLossRate > 0 && Math.random() < link.packetLossRate) {
                this.dropPacket(packet, `Packet lost on link ${link.id} (loss rate ${(link.packetLossRate * 100).toFixed(1)}%)`);
                return;
            }

            packet.currentLink = link;
            packet.currentDevice = fromDev;
            packet.progressOnLink = 0.0;

            if (!this.activeInFlightPackets.includes(packet)) {
                this.activeInFlightPackets.push(packet);
            }

            this.log(`Link Transmission: ${fromDev.name} -> ${toDev.name} over ${link.id} [${link.areaType ? link.areaType.id : 'LINK'}, delay=${link.latencyMs}ms]`);
        }

        deliverPacket(destDev, packet) {
            packet.status = 'DELIVERED';
            packet.currentDevice = destDev;
            packet.currentLink = null;
            packet.progressOnLink = 1.0;
            this.activeInFlightPackets = this.activeInFlightPackets.filter(p => p !== packet);

            const latency = Date.now() - packet.creationTime;
            this.metrics.recordPacketDelivered(latency);
            packet.addHop(destDev.name);

            this.notify('onPacketDelivered', packet);
            this.log(`SUCCESS: Packet ${packet.id} DELIVERED at ${destDev.name} (${destDev.getPrimaryIp()}). Latency: ${latency} ms, Hops: ${packet.hopPath.join(' -> ')}`);

            this.processServerReply(destDev, packet);
        }

        dropPacket(packet, reason) {
            packet.dropReason = reason;
            packet.status = 'DROPPED';
            this.activeInFlightPackets = this.activeInFlightPackets.filter(p => p !== packet);
            this.metrics.recordPacketDropped();

            this.notify('onPacketDropped', packet, reason);
            this.log(`DROP: Packet ${packet.id} DROPPED. Reason: ${reason}`);
        }

        processServerReply(destDev, packet) {
            const protoCode = packet.protocol ? packet.protocol.code : '';
            // ICMP Ping Echo Reply
            if (protoCode === 'ICMP' && packet.payload.includes('Echo Request')) {
                const reply = Packet.createPingReply(packet.destIp, packet.sourceIp, packet.pingSequence);
                reply.sourceMac = destDev.getPrimaryMac();
                setTimeout(() => this.sendPacket(reply), 100);
            }
            // HTTP Web Server Response
            else if (protoCode === 'HTTP' && destDev.getDeviceType() === 'Server' && destDev.httpServiceRunning) {
                if (packet.httpMethod === 'GET') {
                    const page = destDev.getWebPage(packet.httpUrl || '/');
                    const httpReply = Packet.createHttpResponse(packet.destIp, packet.sourceIp, 200, page);
                    setTimeout(() => this.sendPacket(httpReply), 150);
                }
            }
            // DNS Server Response
            else if (protoCode === 'DNS' && destDev.getDeviceType() === 'Server' && destDev.dnsServiceRunning) {
                const domain = packet.dnsQuery;
                const resolved = this.dnsService.resolve(domain);
                if (resolved) {
                    const dnsReply = Packet.createDnsResponse(packet.destIp, packet.sourceIp, domain, resolved);
                    setTimeout(() => this.sendPacket(dnsReply), 80);
                }
            }
        }

        startLoop() {
            // Animation Tick Loop (every 30ms)
            this.simInterval = setInterval(() => {
                if (!this.isRunning || this.isPaused) return;
                this.tickPackets();
            }, 30);

            // Metrics Tick Loop (every 1000ms)
            this.metricsInterval = setInterval(() => {
                this.metrics.tickSample();
                this.notify('onStatsUpdated', this.metrics);
            }, 1000);
        }

        tickPackets() {
            for (const packet of [...this.activeInFlightPackets]) {
                const link = packet.currentLink;
                if (!link) continue;

                const speedFactor = this.simulationSpeed;
                const step = (0.05 * speedFactor) / Math.max(link.latencyMs / 10.0, 0.5);
                const newProgress = packet.progressOnLink + step;

                if (newProgress >= 1.0) {
                    const fromDev = packet.currentDevice;
                    const nextDev = link.getOtherDevice(fromDev);
                    packet.progressOnLink = 1.0;
                    this.activeInFlightPackets = this.activeInFlightPackets.filter(p => p !== packet);

                    if (nextDev) {
                        packet.addHop(nextDev.name);
                        packet.currentDevice = nextDev;
                        this.routeFromDevice(nextDev, packet);
                    } else {
                        this.dropPacket(packet, 'Link leads to nowhere');
                    }
                } else {
                    packet.progressOnLink = newProgress;
                    this.notify('onPacketMoved', packet);
                }
            }
        }

        play() {
            this.isPaused = false;
            this.log('Simulation resumed.');
        }

        pause() {
            this.isPaused = true;
            this.log('Simulation paused.');
        }

        step() {
            if (!this.isPaused) this.pause();
            this.tickPackets();
            this.log('Simulation single-step executed.');
        }

        setSimulationSpeed(speed) {
            this.simulationSpeed = Math.max(0.1, Math.min(10.0, speed));
            this.log(`Simulation speed set to ${this.simulationSpeed.toFixed(1)}x`);
        }

        log(msg) {
            this.notify('onLogMessage', msg);
        }

        notify(event, ...args) {
            for (const listener of this.listeners) {
                if (typeof listener[event] === 'function') {
                    listener[event](...args);
                }
            }
        }

        reset() {
            this.activeInFlightPackets = [];
            this.packetHistory = [];
            this.metrics.reset();
            for (const d of this.devices) {
                d.arpTable.clear();
                if (d.clearMacTable) d.clearMacTable();
            }
            this.recalculateRouting();
            this.log('Simulation reset.');
        }

        shutdown() {
            this.isRunning = false;
            if (this.simInterval) clearInterval(this.simInterval);
            if (this.metricsInterval) clearInterval(this.metricsInterval);
        }
    }

    exports.NetworkSimulator = NetworkSimulator;
})(typeof module !== 'undefined' && module.exports ? module.exports : (window.CampusNet = window.CampusNet || {}));
